src/conversion_helper.js

```diff
@@ -758,34 +780,66 @@ function filterConversions(potentialConversions) {
 /*
   Input: imperial conversions
     [
-      { "imperial" : { "number" : 10000, "unit" : " miles" } },
-      { "imperial" : { "number" : 30, "unit" : " mpg" } }
+      { "imperial" : { "numbers" : 10000, "unit" : " miles" } },
+      { "imperial" : { "numbers" : 30, "unit" : " mpg" } }
     ]
   Output: metric and imperial conversions
     [
       { 
         "imperial" : 
-          { "number" : 10000, "unit" : " miles" }, 
+          { "numbers" : 10000, "unit" : " miles" }, 
         "metric": 
-          { "number" : 16093.44, "unit" : " km" } 
+          { "numbers" : 16093.44, "unit" : " km" } 
       },
       { 
         "imperial" : 
-          { "number" : 30, "unit" : " mpg" }, 
+          { "numbers" : 30, "unit" : " mpg" }, 
         "metric" : [
-          { "number" : 12.7543, "unit" : " km/L" },
-          { "number" : 7.84049, "unit" : " L/100km" } 
+          { "numbers" : 12.7543, "unit" : " km/L" },
+          { "numbers" : 7.84049, "unit" : " L/100km" } 
         ]
       }
     ]
 */
 function calculateMetric(imperialInputs) {
   return imperialInputs.map(input => {
     const imperialUnit = input['imperial']['unit'];
-    const imperialNumber = Number(input['imperial']['number']);
+    const imperialNumbers = input['imperial']['numbers'];
 
     const map = unitLookupMap[imperialUnit];
-    input['metric'] =  map['conversionFunction'](imperialNumber)
+    inRangeVal = -Infinity;
+    let multipleResults = false;
+    imperialNumbers.forEach(function(item) {
+        var tempMap = map['conversionFunction'](Number(item));
+        if(Array.isArray(tempMap)) {
+          multipleResults = true;
+        }
+    });
+
+    if(multipleResults) {
+      input['metric'] = [];
```

Comment:
I think there's probably a way to use `.map` on imperialNumbers instead of using a `.forEach` and then pushing onto the `input['metric']` array. It might make the code clearer

Reply:
I don't think so. If you look closely, I'm manipulating indices inside input['metric'] as well. Even if there is one, I don't think it'll make it clearer. Though, I haven't really used `.map` much before, so I'm probably wrong about this, so can you explain how that would work?

Reply:
### About `map`
Alright, here's why I think `map` is cool:

"map" is a function called on an array and it takes in a function as a parameter:  
`const foo = [1, 2, 3].map(i => i.toString())`  
It applies the function to each element of the array, and then returns back an array `foo` is now `["1", "2", "3"]`.

It's the exact same thing as declaring an empty list `let foo = []`, and doing:  
`[1, 2, 3].forEach(i => foo.push(i.toString()))`  
`foo` is again equal to `["1", "2", "3"]`

Using a `map` instead of forEach/push is a more "functional" way to program, rather than "imperative." "Functional" code involves less changing state (you see how we can declare foo as const in the first example?) and this generally makes code easier to understand.

---

### About the refactor
I think `calculateMetric` can be refactored to:

```
const map = unitLookupMap[imperialUnit];
const conversionFn = map['conversionFunction'];

return imperialInputs.map((imperial) => {
  const imperialNumbers = imperial["numbers"];
  return Object.assign(imperial, {metric: conversionFn(imperialNumbers)});
});
```

but then we would have to refactor conversionFn to:
```
{
  "imperialUnits" : [/psi/, /pounds?[ -]?(?:force)?[- ]?(?:per|an?[/])[- ]?squared? inch/],
  "standardInputUnit" : " psi",
  "isInvalidInput" : isZeroOrNegative,
  "isWeaklyInvalidInput" : isHyperbole,
  "conversionFunction" : (arr) => pressureMap(arr, 6894.76)
}
```

So that would make the pressureMap (<-- and all the other ones)
```
unction pressureMap(imperialNumbers, metricConversion) {
  const metricNumbers = imperialNumbers.map((imperial) => imperial * metricConversion);
  const unitBasedOnNumber = Math.max(metricNumbers);

  if (unitBasedOnNumber < 1000) {
    return createMap(metricNumbers, 1, " Pa");

  } else {
    return createMap(metricNumbers, 1/1000, " kPa");
  }
}
```

So that would make createMap
```
function createMap(valueArray, multiplier, unit) {
  return {
    "numbers" : valueArray.map((val) => (val * multiplier).toString()),
    "unit" : unit
  };
}
```

...basically, instead of "unwrapping" the array and dealing with each individual element at the top-level of the `calculateMetric` function, my idea is to push down the unwrapping to the `createMap` level.

Does that make sense? Any thoughts? I haven't actually compiled this and you're probably more familiar with the code here than I am now, haha.

Reply:
Ah, this is a good idea too. Passing the array all the way down to the functions. This would however probably need a proper refactor of the conversion fn and distancemap etc.

Reply:
Instead of mapping it down within `createMap`, I think we'll have to make separate functions to do math on arrays inline. Because precision. Doing stuff that way, I'm getting weird errors, 92.83000000002 instead of 92.83 which are hard to avoid as we interchange to strings as well. What are your thoughts?

Reply:
I think the floating point errors are not too big of a deal, we are rounding the numbers at the end anyways.

Reply:
And yes, conversion function and all the "distance apart" stuff would need to be refactored. Open to other suggestions though

Reply:
Actually, the main problem would be dealing with tests for calculateMetric, since precision errors will skew the results unpredictably. One solution might be to allow eps difference in expected and found value, but that would complicate it a bit.

Reply:
I think if we have just a couple more functions, multiplyArray and dividebyArray, which do just that, we’d eliminate the need for changing the tests from current state.

Reply:
Oh, I see, floating points make "equals" not work in tests. I'm not a fan of changing production code just so that we can test them better, perhaps there is a "test" solution:

Chai (the assertion library we are using) has a [closeTo](http://chaijs.com/api/bdd/#method_closeto) method, where you can tell it an acceptable range

There is also this [deep closeTo](https://www.npmjs.com/package/chai-deep-closeto) library that does not look very well maintained... but it exists! We can see if still it works or not.

There is an issue on [Chai's Issue page](https://github.com/chaijs/chai/issues/412) talking about this very problem. If "deep closeTo" does not work and we cannot find a library that exists to solve this, maybe we can make a plugin/library that works

Reply:
Well, it’s not just about the tests, I realised ingraining multiplication will also have some unwanted effects in future, take for example [line 142](https://github.com/cannawen/metric_units_reddit_bot/blob/3d01bf8fc0b8b40f72d8c1485320e61ab9246624/src/conversion_helper.js#L142), this can’t be expressed as multiplication. Handling units like this would need special functions for everything, making adding units in future much more complicated. So we might as well write functions that could do that for someone...

Reply:
That is a good point, it's not always multiplication. Sorry if I misunderstood!! Are you talking about something like this?

```
{
  "imperialUnits" : [/psi/, /pounds?[ -]?(?:force)?[- ]?(?:per|an?[/])[- ]?squared? inch/],
  "standardInputUnit" : " psi",
  "isInvalidInput" : isZeroOrNegative,
  "isWeaklyInvalidInput" : isHyperbole,
  "conversionFunction" : (imperialNumbers) => pressureMap(imperialNumbers, (i) -> i * 6894.76)
}
```
Passing in a function, instead of a number?

Reply:
Yep. And for conversionFn’s calling createMap directly, just use the .map method inline.

Reply:
But note that this will remove the flexibility for changing the final unit based on the value(it will have to be max for all conversionFns)

Reply:
I think "max" is probably fine for all of them for now, but we can always re-visit and refactor if we see that it is a problem in the future

Reply:
Finally done with these changes now. Thanks for all the help, and sorry I took so much time to understand so much stuff